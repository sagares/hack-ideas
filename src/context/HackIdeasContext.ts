import { createContext } from "react";
import { IdeasArray } from "../HackIdea.types";

type Employee = {
  empId: string;
};

type HackIdeasContextType = {
  currentUser: Employee;
  logout: () => void;
  setCurrentUser: (employee: Employee) => void;
  ideas: IdeasArray;
  setIdeas: (ideas: IdeasArray) => void;
};

const HackIdeasContext = createContext<Partial<HackIdeasContextType>>({});

export { HackIdeasContext };
