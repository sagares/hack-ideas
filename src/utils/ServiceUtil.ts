// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getDatabase, ref, onValue, set } from "firebase/database";
import {  Idea } from "../HackIdea.types";

const firebaseConfig = {
  apiKey: "AIzaSyC0RFn0RNnX8EoPHpAu-lf4EvN08aPKMCg",
  authDomain: "hack-ideas-c7cd7.firebaseapp.com",
  databaseURL: "https://hack-ideas-c7cd7-default-rtdb.firebaseio.com",
  projectId: "hack-ideas-c7cd7",
  storageBucket: "hack-ideas-c7cd7.appspot.com",
  messagingSenderId: "197788164361",
  appId: "1:197788164361:web:36a94bf614b4b5c830f3d8",
  measurementId: "G-660ZLLJJEB",
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const database = getDatabase();

const registerEmployee = (empId: string) => {
  set(ref(database, "employees/" + empId), {
    empId,
  });
};

const getEmployeeById = (empId: string, callback: Function) => {
  let empRef = ref(database, "employees/" + empId);
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getEmployeeData = (callback: Function) => {
  let empRef = ref(database, "employees/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getIdeas = (callback: Function) => {
  let empRef = ref(database, "ideas/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const registerIdea = (idea: Idea) => {
  const id = toCamelCase(idea.name);
  set(ref(database, "ideas/" + id), {
    ...idea,
    id,
    upvotes: [],
  });
};

const updateIdea = (ideaId: string, updatedIdea: Idea) => {
  set(ref(database, "ideas/" + ideaId), updatedIdea);
};

const toCamelCase = (sentenceCase: string) => {
  var out = "";
  sentenceCase.split(" ").forEach((el, idx) => {
    var add = el.toLowerCase();
    out += idx === 0 ? add : add[0].toUpperCase() + add.slice(1);
  });
  return out;
};

export {
  getEmployeeById,
  getEmployeeData,
  getIdeas,
  registerEmployee,
  registerIdea,
  updateIdea,
};
