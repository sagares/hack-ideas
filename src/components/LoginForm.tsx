import React, { useContext, useEffect, useState } from "react";
import M from "materialize-css";
import HackIdeasContext from "../context";
import { useSessionStorage } from "../hooks";
import useFormInput from "../hooks/useFormInput";
import { getEmployeeById, registerEmployee } from "../utils/ServiceUtil";

import "./login-form.less";

const LoginForm = () => {
  const { setCurrentUser } = useContext(HackIdeasContext);
  const [userInSession, setUserInSession] = useSessionStorage("employee", null);
  const employeeId = useFormInput("");
  const [isValid, setIsValid] = useState(true);

  const handleChange = (e: React.FormEvent<HTMLInputElement>) => {
    const re = /^[0-9\b]{1,6}$/;
    if (e.currentTarget.value === "" || re.test(e.currentTarget.value)) {
      employeeId.onChange(e);
    }
  };

  const handleLogin = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const re = /^[0-9\b]{6}$/;
    if (re.test(employeeId.value)) {
      setIsValid(true);
      proceedToLogin();
    } else {
      setIsValid(false);
    }
  };

  const proceedToLogin = () => {
    getEmployeeById(employeeId.value, (employee) => {
      if (employee) {
        setCurrentUser(employee);
        setUserInSession(employee);
      } else {
        registerEmployee(employeeId.value);
        setCurrentUser({ empId: employeeId.value });
        setUserInSession({ empId: employeeId.value });
      }
    });
  };

  useEffect(() => {
    setCurrentUser(userInSession);
    const tooltips = document.querySelectorAll('.tooltipped');
    M.Tooltip.init(tooltips, {});
  }, []);

  return (
    <div className="card-panel login-form-card z-depth-3">
      <div className="row valign-wrapper login-form-wrapper">
        <div className="col s12 m12 l6">
          <h2>{`{ Hack Ideas }`}</h2>
          <h6>Enter With Employee Id</h6>
        </div>
        <form className="col s12 m12 l6" noValidate onSubmit={handleLogin}>
          <div className="row valign-wrapper">
            <div className="input-field col s12">
              <input
                id="empId"
                type="text"
                className="validate tooltipped"
                data-position="bottom"
                data-tooltip="Enter your 6 digit employee ID"
                value={employeeId.value}
                onChange={handleChange}
                data-error="Employee ID is required"
                required
              />
              <label htmlFor="empId">Employee ID</label>
              {!isValid && (
                <span className="helper-text" data-error="Enter valid employee id.">
                  Enter valid employee id.
                </span>
              )}
            </div>
          </div>
          <div className="row">
            <button
              type="submit"
              name="btn_login"
              className="col s6 m12 l12 btn btn-large waves-effect"
              disabled={!employeeId.value}
            >
              Login
              <i className="material-icons right">send</i>
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

export default LoginForm;
