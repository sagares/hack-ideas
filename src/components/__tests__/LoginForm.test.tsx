/**
 * @jest-environment jsdom
 */
import React from "react";
import { act, fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { getEmployeeById, registerEmployee } from "../../utils/ServiceUtil";
import HackIdeasContext from "../../context";
import LoginForm from "../LoginForm";


jest.mock("../../utils/ServiceUtil");
jest.mock("../login-form.less", () => jest.fn());

describe("Login From Component", () => {
  let element, component, setCurrentUser;

  beforeEach(() => {
    (getEmployeeById as jest.Mock).mockImplementation((employee, cb) => {
      cb({empId: employee});
      return jest.fn();
    });

    (registerEmployee as jest.Mock).mockReturnValue({});

    setCurrentUser = jest.fn().mockImplementation(() => {});

    element = document.createElement("div");
    component = render(
      <HackIdeasContext.Provider value={{ setCurrentUser }}>
          <LoginForm />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );
  });

  afterEach(() => {
    document.body.removeChild(element);
    jest.clearAllMocks();
  });

  test("renders login form", () => {
    expect(component.container).toMatchSnapshot();
  });

  test("updates input value on change event and enabled login button", () => {
    const empIdInput = component.container.querySelector("#empId");
    const loginBtn = component.container.querySelector("button[name=\"btn_login\"]");

    fireEvent.change(empIdInput, { target: { value: "123456" } });
    expect(empIdInput.value).toBe("123456");
    expect(loginBtn.disabled).toBe(false);
  });

  test("updates input value if value is empty string and disables login button", () => {
    const empIdInput = component.container.querySelector("#empId");
    const loginBtn = component.container.querySelector("button[name=\"btn_login\"]");

    fireEvent.change(empIdInput, { target: { value: "" } });
    expect(empIdInput.value).toBe("");
    expect(loginBtn.disabled).toBe(true);
  });

  test("does not update input if value is not number", () => {
    const empIdInput = component.container.querySelector("#empId");

    fireEvent.change(empIdInput, { target: { value: "123" } });
    expect(empIdInput.value).toBe("123");

    fireEvent.change(empIdInput, { target: { value: "ab" } });
    expect(empIdInput.value).toBe("123");
  });

  test("does not update input if value is more than 6 digit number", () => {
    const empIdInput = component.container.querySelector("#empId");

    fireEvent.change(empIdInput, { target: { value: "123" } });
    expect(empIdInput.value).toBe("123");

    fireEvent.change(empIdInput, { target: { value: "1234567" } });
    expect(empIdInput.value).toBe("123");
  });

  test("displays error message on login when employee id is not valid", () => {
    const empIdInput = component.container.querySelector("#empId");
    const loginBtn = component.container.querySelector("button[name=\"btn_login\"]");

    fireEvent.change(empIdInput, { target: { value: "123" } });
    expect(empIdInput.value).toBe("123");

    act(() => {
      loginBtn.click(loginBtn);
    });
    
    expect(component.container.querySelector(".helper-text")).toBeDefined();
  });

  test("does not display error message on login when employee id is valid", () => {
    const empIdInput = component.container.querySelector("#empId");
    const loginBtn = component.container.querySelector("button[name=\"btn_login\"]");

    fireEvent.change(empIdInput, { target: { value: "123456" } });

    act(() => {
      loginBtn.click(loginBtn);
    });
    
    expect(component.container.querySelector(".helper-text")).toBe(null);
  });

  test("sets current user if the employee is already present", () => {
    const empIdInput = component.container.querySelector("#empId");
    const loginBtn = component.container.querySelector("button[name=\"btn_login\"]");

    fireEvent.change(empIdInput, { target: { value: "123456" } });

    act(() => {
      loginBtn.click(loginBtn);
    });
    
    expect(setCurrentUser).toHaveBeenCalledWith({"empId": "123456"});
  });

  test("registers employee first and then set current user if the employee is not present", () => {
    (getEmployeeById as jest.Mock).mockImplementation((employee, cb) => {
      cb(null);
      return jest.fn();
    });

    const empIdInput = component.container.querySelector("#empId");
    const loginBtn = component.container.querySelector("button[name=\"btn_login\"]");

    fireEvent.change(empIdInput, { target: { value: "550324" } });

    act(() => {
      loginBtn.click(loginBtn);
    });
    
    expect(setCurrentUser).toHaveBeenCalledWith({"empId": "550324"});
  });
});
