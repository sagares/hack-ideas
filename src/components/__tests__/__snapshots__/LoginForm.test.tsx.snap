// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login From Component renders login form 1`] = `
<div>
  <div
    class="card-panel login-form-card z-depth-3"
  >
    <div
      class="row valign-wrapper login-form-wrapper"
    >
      <div
        class="col s12 m12 l6"
      >
        <h2>
          { Hack Ideas }
        </h2>
        <h6>
          Enter With Employee Id
        </h6>
      </div>
      <form
        class="col s12 m12 l6"
        novalidate=""
      >
        <div
          class="row valign-wrapper"
        >
          <div
            class="input-field col s12"
          >
            <input
              class="validate tooltipped"
              data-error="Employee ID is required"
              data-position="bottom"
              data-tooltip="Enter your 6 digit employee ID"
              id="empId"
              required=""
              type="text"
              value=""
            />
            <label
              for="empId"
            >
              Employee ID
            </label>
          </div>
        </div>
        <div
          class="row"
        >
          <button
            class="col s6 m12 l12 btn btn-large waves-effect"
            disabled=""
            name="btn_login"
            type="submit"
          >
            Login
            <i
              class="material-icons right"
            >
              send
            </i>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
